### PR TITLE
NF4 quantization of linear layers without LoRA applied

### PR DESCRIPTION
**Pull request by winglian**
#### Context
What is the purpose of this PR? Is it to
- [X ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses. #1093 

#### Changelog
Reverts #658 to bring back FrozenNF4Linear. When `quantize_base` is set to true, all base weights for linear layers are quantized, even if they do not have LoRA applied to them.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Comment by joecummings** [comment]
cc @msaroufim

**Review by janeyx99 — COMMENTED** [review]
Clarification question: does thid mean that `quantize_base` is the boolean with which one specifies the q in qLoRA, which is globally applied to all projs? (versus how LoRA could be applied to a subset of the parameters of q, k, v and not o).

**Comment by winglian** [comment]
running `tune run lora_finetune_single_device --config llama3/8B_qlora_single_device`
but only with `lora_attn_modules: ['q_proj']` instead of all the attention linear layers, uses `15810MiB` on main and `12688MiB` on this branch.

File: torchtune/modules/low_precision/nf4_linear.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    ):
+        if "bias" in kwargs and kwargs.pop("bias"):
+            raise RuntimeError("FrozenNF4Linear does not currently support biases!")
+
+        super().__init__(in_dim, out_dim, device=device, bias=False, **kwargs)
+        self.weight.requires_grad_(False)
+        self.nf4_weight = to_nf4(self.weight.data)
+        # re-register self.weight as the nf4 weight, so that the nf4 weight
+        # shows up as expected in .parameters, state_dict, etc.
+        self.weight = torch.nn.Parameter(self.nf4_weight, requires_grad=False)
```
Comment (janeyx99) [inline torchtune/modules/low_precision/nf4_linear.py:46]: Could we avoid accessing `.data` of `self.weight`?

I'm curious if there's a way to use swap_tensors here to accomplish this more cleanly. https://pytorch.org/docs/stable/generated/torch.utils.swap_tensors.html  cc @mikaylagawarecki  any ideas?

**Review by janeyx99 — COMMENTED** [review]
Rest of PR looks fine; we should implement FrozenNF4Linear as cleanly as possible though

**Review by janeyx99 — COMMENTED** [review]
Also, what test cases can we write to verify functionality?

File: torchtune/modules/low_precision/nf4_linear.py (hunk 1 above)
Comment (mikaylagawarecki) [reply torchtune/modules/low_precision/nf4_linear.py:46]: @winglian Could you explain the rationale for the `.data` access?

Comment (ebsmothers) [reply torchtune/modules/low_precision/nf4_linear.py:46]: Yeah can we do something similar to what's currently done in [LoRALinear](https://github.com/pytorch/torchtune/blob/58d58bce65836c23865e6a9aa683475631527d82/torchtune/modules/peft/lora.py#L94) for this?

**Review by ebsmothers — COMMENTED** [review]
Overall the changes look reasonable to me. Re @janeyx99's testing comment, a couple things I think we should add here:

(1) a unit test for nf4_linear
(2) some kind of e2e test. I am thinking: run one of our QLoRA recipes before and after the change, confirm that (a) the peak memory reduces as expected, and (b) we see no regression in eval metrics on the resulting fine-tuned checkpoint.

It's also likely that this will break some of our existing QLoRA tests. So e.g. the values [here](https://github.com/pytorch/torchtune/blob/58d58bce65836c23865e6a9aa683475631527d82/tests/recipes/test_lora_finetune_single_device.py#L55-L58) would need to be updated

File: torchtune/modules/low_precision/nf4_linear.py (hunk 1 above)
Comment (winglian) [reply torchtune/modules/low_precision/nf4_linear.py]: This is simply reverting back the code that was deleted in https://github.com/pytorch/torchtune/pull/658/files#diff-74b0d911936ebd1d0e216004577afed27b84b6bfdff9c6a9a1a28f6fac054850L45

Removing the `.data` does seem to work as well, so I've checked in that change.

File: torchtune/modules/low_precision/nf4_linear.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    def __init__(
+        self, in_dim: int, out_dim: int, device: Optional[torch.device] = None, **kwargs
+    ):
+        if "bias" in kwargs and kwargs.pop("bias"):
+            raise RuntimeError("FrozenNF4Linear does not currently support biases!")
+
+        super().__init__(in_dim, out_dim, device=device, bias=False, **kwargs)
+        self.weight.requires_grad_(False)
+        self.nf4_weight = to_nf4(self.weight)
```
Comment (janeyx99) [inline torchtune/modules/low_precision/nf4_linear.py:43]: nice

File: tests/torchtune/modules/low_precision/test_nf4_linear.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        nf4_linear = FrozenNF4Linear(dim, dim, device="cuda", dtype=dtype)
+        orig_weight = nf4_linear.weight.get_original_weight().clone().detach()
+        bnb_nf4_linear = _build_bnb_linear(input_weight=orig_weight)
+
+        # From https://github.com/drisspg/transformer_nuggets/blob/f05afad68ad9086d342268f46a7f344617a02314/test/test_qlora.py#L65
+        bnb_reconstruction = bnb_nf4_linear(
+            torch.eye(dim, dim, dtype=dtype, device="cuda")
+        )
+        # Ensure nf4_linear and bnb reconstructions are close to each other.
+        diff = (
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply tests/torchtune/modules/low_precision/test_nf4_linear.py:100]: and agreed with @ebsmothers  that there should be a test to ensure memory is as expected, either in the code using torch.cuda.max_memory_allocated or with a local e2e test summarized in the PR body.

File: torchtune/modules/low_precision/nf4_linear.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        # re-register self.weight as the nf4 weight, so that the nf4 weight
+        # shows up as expected in .parameters, state_dict, etc.
+        self.weight = torch.nn.Parameter(self.nf4_weight, requires_grad=False)
```
Comment (janeyx99) [inline torchtune/modules/low_precision/nf4_linear.py:46]:
```suggestion
        torch.utils.swap_tensors(self.weight, torch.nn.Parameter(self.nf4_weight, requires_grad=False))
```
Not a big expert of swap_tensor, but something like this should ensure that the non-data portions of self.weight are preserved + all other references to it are preserved.

File: tests/torchtune/modules/low_precision/test_nf4_linear.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+from torchtune.modules.low_precision import FrozenNF4Linear
+from torchtune.utils.seed import set_seed
+
+try:
+    from torchao.dtypes.nf4tensor import NF4Tensor
+except ImportError as e:
+    raise RuntimeError(
+        "Please install torchao to run this test."
+        "Example: pip install git+https://github.com/pytorch-labs/ao.git"
+    ) from e
```
Comment (ebsmothers) [inline tests/torchtune/modules/low_precision/test_nf4_linear.py:19]: Do we still need this?

File: tests/torchtune/modules/low_precision/test_nf4_linear.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ebsmothers) [reply tests/torchtune/modules/low_precision/test_nf4_linear.py:100]: Yeah I also realized @winglian had already done this but only after my previous comment. I think in that case it should be fine to just document it in the PR summary with the command used to get those numbers.

**Review by ebsmothers — COMMENTED** [review]
One other thing -- this will change the default behavior of recipes using `quantize_base`. Can you run some QLoRA fine-tunes and report eval numbers before and after the change to confirm it doesn't cause any model quality regressions? (E.g. if I only apply LoRA to `q_proj` and `v_proj` and set `quantize_base=True` I will now be converting a lot more linears into lower precision than I did before. I don't think this should ultimately matter much but would be good to document that in the PR summary)

File: tests/torchtune/modules/low_precision/test_nf4_linear.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (winglian) [reply tests/torchtune/modules/low_precision/test_nf4_linear.py:93]: yeah, that was using `tune run lora_finetune_single_device --config llama3/8B_qlora_single_device`, with the change of `lora_attn_modules: ['q_proj']` so that we have some linear layers left over to quantize that aren't LoRALinear.

**Comment by winglian** [comment]
here's the evals. slight drop from quantization `0.5682` -> `0.5546`

```
main

|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|------|---|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |↑  |0.5682|±  |0.0149|

nf4 - quantized linear
|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|------|---|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |↑  |0.5546|±  |0.0149|
```

**Comment by ebsmothers** [comment]
@winglian thanks for adding the eval results, I think it looks reasonable. One remaining thing before we merge: seems like one of the bnb comparison unit test cases is failing. See [here](https://github.com/pytorch/torchtune/actions/runs/9878727250/job/27283298692?pr=1119)

**Review by joecummings — APPROVED** [review]
WOoooooooooooooooooooooooooo
